libtiledbsoma/src/reindexer/reindexer.cc

```diff
@@ -0,0 +1,202 @@
+/**
+ * @file   reindexer.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the a pybind11 api into SOMA C++ library.
+ */
+
+#include "reindexer.h"
+#include <unistd.h>
+#include "khash.h"
+
+uint64_t current_time() {
+    uint64_t microseconds_since_epoch =
+        std::chrono::duration_cast<std::chrono::microseconds>(
+            std::chrono::system_clock::now().time_since_epoch())
+            .count();
+    return microseconds_since_epoch;
+}
+
+static std::mutex log_mutex;
+
+static bool no_logging_s = true;
+
+static void log(std::string message, int ln, std::string severity = "INFO") {
+    if (!no_logging_s) {
+        std::lock_guard l(log_mutex);
+        std::cout << current_time() << " LN " << std::to_string(ln) << " LOG("
+                  << severity << "): " << message << std::endl;
+    }
+}
+
+void IntIndexer::map_locations(const int64_t* keys, int size, int threads) {
+    log("Start of Map locations [" + std::to_string(long((void*)keys)) +
+            "] of size " + std::to_string(size) + " and " +
+            std::to_string(threads) + " threads ",
+        __LINE__);
+    hash_ = kh_init(m64);
+    kh_resize(m64, hash_, size * 1.25);
+    log("Thread pool started and hash table created ", __LINE__);
+    int ret;
+    khint64_t k;
+    int64_t counter = 0;
+    // Hash map construction
+    for (int i = 0; i < size; i++) {
+        k = kh_put(
+            m64,
+            hash_,
+            keys[i],
+            &ret);  // hash_wrapper<IntType>::put(hash_, key, &ret);
+        assert(k != kh_end(hash_));
+        kh_val(hash_, k) = counter;
+        counter++;
+    }
+    size_ = size;
```

Comment:
In order to comply with C++ Core Guidelines, class level attributes should be set in the constructor and not inside the function. Lets avoid setting size, thread_count and declaring the thread_pool in this function.